keps/sig-apps/3715-elastic-indexed-job/README.md

```diff
@@ -0,0 +1,930 @@
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+I have a batch workload that requires two sets of pods: a manager and a group of workers that can scale up or down. 
+The manager communicates with the workers to manage their load; examples of such workloads include distributed HPC 
+and districuted ML training. Examples of frameworks include Horovod with MPI, Spark and Ray.
+
+This workload can be modeled as two `Indexed` Jobs, one for the manager and one for the workers. A 
+headless service is created to set up stable hostnames for the worker pods . The workers job is
+scaled up/down by updating `spec.completions` and `spec.parallelism` in tandem.
+
+The success semantics of these workloads are tied to the manager, not the workers. However, in
+such workloads, when the job is finished the workers are either deleted directly (effectively scaled down to 0)
+or they potentially exit on their own after communicating their final status to the manager who decides the final status
+of the whole workload.
+
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+There is a concern that changing completion semantics on the fly can be subject to
+race conditions with declaring the job finished. However, after a closer look, this 
+seems to not be an issue because the job controller uses Update when setting the
+finished status on the job object, which is subject to object modified errors if the spec
+changes at the same time.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+
+The controller already counts the number of completed indexes within the range defined by `spec.completions` and
+uses that to evaluate if the job completed successfully and to set the `status.Succeeded` count. This is not impacted
+if `spec.completions` changes over time. Morever, on scale down, the controller also automatically removes any indexes
+outside the range. In job stat
+
+
+As mentioned above, pod failures may have happened to indexes that are outside the range after `spec.completions`
+is updated. Those failures will still count against the backoffLimit of the job and the field `status.Failed` will
+not be decremented.
+
```

Comment:
@liggitt @soltysh since the change is only relaxing validation, I suggest we start the feature in beta and enabled by default right away. Just like we successfully did for https://github.com/kubernetes/enhancements/issues/2926.

Any objections?

Reply:
> The next question to ask is what a 1.26-level controller would do with a job that mutated its completions/parallelism.

It should handle it correctly. In fact, as of writing this KEP, 1.27 and 1.26 are identical.

If tests reveal that we need to make changes rendering the above not valid, then we will limit it to alpha.

Reply:
> It should handle it correctly. In fact, as of writing this KEP, 1.27 and 1.26 are identical.

that's great to hear

> If tests reveal that we need to make changes rendering the above not valid, then we will limit it to alpha.

sgtm

Reply:
Do we already have a log that trims the `CompletedIndexes` field appropriately? Are we just computing it from scratch every time?

Reply:
It gets re-calculated on every sync and stored in status if it changes.

Reply:
If by computing it from scratch you mean that we look at all the existing pods, the answer is NO. We fixed that with tracking with finalizers :smiley: 

But we do parse the field, and update accordingly, considering the newly finished pods.